typhos/display.py

```diff
@@ -1102,29 +1138,90 @@ def _move_display_to_layout(self, widget):
+        ]
+
+    def _generate_template_menu(self, base_menu: QtWidgets.QMenu) -> None:
         """Generate the template switcher menu, adding it to ``base_menu``."""
-        for view, filenames in self.templates.items():
-            if view.endswith('_screen'):
-                view = view.split('_screen')[0]
-            menu = base_menu.addMenu(view.capitalize())
+        dev = self.device
+        if dev is None:
+            return
 
-            for filename in filenames:
-                def switch_template(*, filename=filename):
-                    self.force_template = filename
+        actions: List[QtWidgets.QAction] = []
+
+        def add_template(filename: pathlib.Path) -> None:
+            def switch_template(*, filename: pathlib.Path = filename):
+                self.force_template = filename
+
+            action = base_menu.addAction(str(filename))
+            action.triggered.connect(switch_template)
+            actions.append(action)
+
+            if self.current_template == filename:
+                base_menu.setDefaultAction(action)
+
+        def add_header(label: str, icon: Optional[QtGui.QIcon] = None) -> None:
+            action = QtWidgets.QWidgetAction(base_menu)
+            label = QtWidgets.QLabel(label)
+            label.setObjectName("menu_template_section")
+            action.setDefaultWidget(label)
+            if icon is not None:
+                action.setIcon(icon)
+            base_menu.addAction(action)
+
+        self._refresh_templates()
+        seen = set()
+
+        for template_type in DisplayTypes:
+            added_header = False
+            for cls in type(dev).mro():
+                matching = self._get_matching_templates_for_class(cls, template_type)
+                templates = set(matching) - seen
+                if not templates:
+                    continue
 
-                action = menu.addAction(os.path.split(filename)[-1])
-                action.triggered.connect(switch_template)
+                def by_match_order(template: pathlib.Path) -> int:
+                    return matching.index(template)
 
-        refresh_action = base_menu.addAction("Refresh Templates")
-        refresh_action.triggered.connect(self._refresh_templates)
+                if not added_header:
+                    add_header(
+                        f"{template_type.friendly_name} screens",
+                        icon=TyphosToolButton.get_icon(
+                            TyphosDisplaySwitcherButton.icons[template_type.name]
+                        ),
+                    )
+                    added_header = True
+
+                base_menu.addSection(f"{cls.__name__}")
+                for filename in sorted(templates, key=by_match_order):
+                    add_template(filename)
+
+        add_header("Typhos default screens")
+        for template in DEFAULT_TEMPLATES_FLATTEN:
+            add_template(template)
+
+        prefix = os.path.commonprefix(
```

Comment:
Why is this widely useful utility function hiding in the os.path module? Python weird

Reply:
I couldn't agree more...